test/integration2/migrate/migration-lifecycle-integration.spec.js

```diff
@@ -0,0 +1,317 @@
+            expect(afterEach.callCount).to.equal(2);
+            expect(tableExistenceChecks).to.deep.equal([
+              {
+                hasFirstTestTable: true,
+                hasSecondTestTable: false,
+              },
+              {
+                hasFirstTestTable: true,
+                hasSecondTestTable: true,
+              },
+            ]);
+          });
+
+          it('is not called after a migration fails', async function () {
+            const afterEach = sinon.stub();
+
+            await knex.migrate
+              .latest({
+                directory: 'test/integration2/migrate/test_with_invalid',
+                afterEach,
+              })
+              .catch(() => {});
+
+            // The afterEach hook should have run for the first two successful migrations, but
+            // not after failed third migration
+            expect(afterEach.callCount).to.equal(2);
+            expect(
+              afterEach.args.map(([_knex, args]) => {
+                return args.file;
+              })
+            ).to.deep.equal([
+              '20131019235242_migration_1.js',
+              '20131019235306_migration_2.js',
+            ]);
+          });
+
+          it('does not run the afterAll hook if the hook fails', async function () {
+            const afterEach = sinon
+              .stub()
+              .throws(new Error('force afterEach hook failure'));
+            const afterAll = sinon.stub();
+
+            await knex.migrate
+              .latest({
+                directory: 'test/integration2/migrate/test',
+                afterEach,
+                afterAll,
+              })
+              .catch((error) => {
+                expect(error.message).to.equal('force afterEach hook failure');
+              });
+
+            expect(afterAll.called).to.be.false;
+          });
+        });
+
+        describe('execution order', function () {
+          it('runs in the expected order of beforeAll -> beforeEach -> afterEach -> afterAll', async function () {
+            const order = [];
+
+            await knex.migrate.latest({
+              directory: 'test/integration2/migrate/test',
+              beforeAll: () => order.push('beforeAll'),
+              beforeEach: (_knex, { file }) => order.push(`beforeEach-${file}`),
+              afterEach: (_knex, { file }) => order.push(`afterEach-${file}`),
+              afterAll: () => order.push('afterAll'),
+            });
+
+            expect(order).to.deep.equal([
+              'beforeAll',
+              'beforeEach-20131019235242_migration_1.js',
+              'afterEach-20131019235242_migration_1.js',
+              'beforeEach-20131019235306_migration_2.js',
+              'afterEach-20131019235306_migration_2.js',
+              'afterAll',
+            ]);
+          });
+        });
+
+        // NOTE (PR #5541) This is disabled pending further discussion
+        describe.skip('when there are no pending migrations', function () {
```

Comment:
This test is currently failing because there appears to be no short-circuiting check on the length of new migrations returned by `getNewMigrations`. https://github.com/knex/knex/blob/e2bc612179a8971d167bbb0d571f51fe4e602eac/lib/migrations/migrate/Migrator.js#L74

Is there a specific reason we don't short-circuit, or was this an oversight?

Reply:
I think without the hooks it did not matter, but now it does... you should add short circus

Reply:
Done!

Reply:
Done! I've also relocated the `beforeAll`/`afterAll` hook calls to a place that's more DRY.